docs/Coding_Style.md

```diff
@@ -0,0 +1,148 @@
+    LogMessageKeys.PRIMARY_KEY, record.getPrimaryKey(),
+    LogMessageKeys.RECORD_TYPE, record.getRecordType());
+```
+
+By structuring our exceptions this way, we get a number of nice benefits:
+
+* We can easily search our logs for all occurrances of a given exception, 
+  such as finding how often "Found corrupt record" occurs
+* For a given exception, we can filter or group on the logging properties,
+  such as counting the number of correupt records by record type
+* At a broader scale, without having variables polluting the text of the
+  exception, we can trivially do things like determine which exception
+  happens most frequently in the system, by performing a count grouped by
+  the exception title. 
+
+## Logging
+
+Logging in record layer follows a similar paradigm to that described in 
+**Exception structure**, above.  Namely:
+
+* The value logged should always be a `KeyValueLogMessage`
+* The text of `KeyValueLogMessage` must always be a static string, with no variables
+* Logging properties should be added to give context to the error (where 
+  variables would have been used)
+* The logging properties should provide as much detail as possible to diagnose the 
+  source of the error
+
+For example, instead of logging:
+
+```
+LOGGER.info("Unable to open file: " + filename);
+```
+
+you should instead log:
+
+```
+LOGGER.info(KeyValueLogMessage.of("Unable to open file",
+    LogMessageKeys.FILENAME, filename));
+```
+
+## Futures
+
+### `join()` or `get()` in production code
+
+Production code (i.e. code that aren't tests), should never call `CompletableFuture.join()` 
+or `CompletableFuture.get()`, but instead should call `FDBRecordContext.asyncToSync()` or
+`FDBDatabase.asyncToSync()`.  The `asyncToSync()` method provides a number of important
+benefits:
+
+* It establishes a client-configurable configurable time limit on how long we are 
+  willing to wait for the operation to complete
+* It automatically defines a wait timer that will be included in the `FDBStoreTimer`
+  to track the time spent waiting on the event
+* It takes care of error handling and wrapping to promote thrown exceptions to the 
+  appropriate `RecordCoreException`
+* It has the ability to detect if `asyncToSync()` is being called from within a 
+  completing future (see **Blocking in asyncronous code**, below)
+
+### Blocking in asynchronous code
+
+The methods `join()`, `get()`, `asyncToSync()` or any method that is known to block 
+(such as syncronous I/O methods), just never be called from within a completable 
+future.
+
+The thread pool `Executor` has a limited number of threads that are used to invoke 
+the lambda in the `CompletableFuture` when it completes, so if that completable
+makes a blocking call, it has permenantly consumed the `Executor` thread until 
+the completion of that operation. In the case where the blocking call is waiting
+for another future to complete (e.g. `asyncToSync()`, `join()` or `get()`) you
```

Comment:
```suggestion
for another future to complete (e.g., `asyncToSync()`, `join()`, or `get()`), you
```

I suppose the Oxford comma is a bit of a style thing. Maybe we need a comma style document explaining our stance on it 🙃